File: apps/server_py/app/domain/session/write_session_file.py
from __future__ import annotations

import asyncio
from pathlib import Path

from app.domain.session.session_paths import SESSIONS_DIR
from app.domain.session.types import Session

_locks: dict[str, asyncio.Lock] = {}


def session_file_path(session_id: str) -> Path:
    return SESSIONS_DIR / f"{session_id}.json"


def _lock_for(session_id: str) -> asyncio.Lock:
    lock = _locks.get(session_id)
    if lock is None:
        lock = asyncio.Lock()
        _locks[session_id] = lock
    return lock


async def write_session_file(session_id: str, session: Session) -> None:
    async with _lock_for(session_id):
        await asyncio.to_thread(SESSIONS_DIR.mkdir, parents=True, exist_ok=True)
        final_path = session_file_path(session_id)
        tmp_path = final_path.with_suffix(".json.tmp")
        payload = session.model_dump_json()
        await asyncio.to_thread(tmp_path.write_text, payload, encoding="utf-8")
        await asyncio.to_thread(tmp_path.replace, final_path)

File: apps/server/src/models/session/writeSessionFile.ts
import { mkdir, rename, writeFile } from 'fs/promises';
import { join } from 'path';
import { SESSION_DATA_DIR } from './sessionPaths.js';
import type { Session } from './types.js';

export const sessionFilePath = (sessionId: string): string =>
  join(SESSION_DATA_DIR, `${sessionId}.json`);

const sessionTempFilePath = (sessionId: string): string =>
  join(SESSION_DATA_DIR, `${sessionId}.json.tmp`);

const writeQueues = new Map<string, Promise<void>>();

/** 同 session 串行落盘；先写 tmp 再 rename，避免并发/中断导致 JSON 损坏。 */
export const writeSessionFile = async (sessionId: string, session: Session): Promise<void> => {
  const previous = writeQueues.get(sessionId) ?? Promise.resolve();

  const current = previous
    .then(async () => {
      await mkdir(SESSION_DATA_DIR, { recursive: true });
      const tmpPath = sessionTempFilePath(sessionId);
      const finalPath = sessionFilePath(sessionId);
      await writeFile(tmpPath, JSON.stringify(session), 'utf-8');
      await rename(tmpPath, finalPath);
    })
    .finally(() => {
      if (writeQueues.get(sessionId) === current) {
        writeQueues.delete(sessionId);
      }
    });

  writeQueues.set(sessionId, current);
  await current;
};
